### PR TITLE
automatic official FASTA file fetching, several new utility functions related to structure including flexible 3d alignment that supports different length chains to be aligned!

### PR DESCRIPTION
**Pull request by YoelShoshan**

### CODE REVIEW

File: fusedrug/data/protein/structure/extract_chains_to_pdbs.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+from jsonargparse import CLI
+from fusedrug.data.protein.structure.structure_io import (
+    load_pdb_chain_features,
+    save_structure_file,
+)
+from typing import Optional
+
+
+def main(
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/extract_chains_to_pdbs.py:9]: what is this script for? please explain in the comments section.

File: fusedrug/data/protein/structure/flexible_align_chains_structure.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        },
+        chain_to_aa_index_seq={
+            chain_id: apply_rigid_on_dynamic_chains[chain_id]["aatype"]
+            for chain_id in apply_rigid_on_dynamic_chains.keys()
+        },
+        save_cif=False,
+        mask=None,  # TODO: check
+    )
+
+    # apply_on_atom_pos = apply_rigid_on_dynamic_concat['atom_positions']
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/flexible_align_chains_structure.py:176]: Do you need the commented out code below?

File: fusedrug/data/protein/structure/protein_complex.py [hunk 3]
```diff
@@ -22,18 +22,26 @@ def __init__(self, verbose: bool = True) -> None:
         self.chains_data = {}  # maps from chain description (e.g. ('7vux', 'A')) to
         self.flattened_data = {}
 
+        #
+        self.per_chain_most_frequent_residue_part = {}
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/protein_complex.py:26]: The key is chain_id?

File: fusedrug/data/protein/structure/protein_complex.py [hunk 4]
```diff
@@ -22,18 +22,26 @@ def __init__(self, verbose: bool = True) -> None:
...
+        self.per_chain_mmcif_object = {}
+        self.per_chain_mmcif_dict = {}
+
     def add(
         self,
-        pdb_id: str,
+        pdb_id_or_filename: str,
+        pdb_id: Optional[str] = None,
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/protein_complex.py:33]: Why do you keep pdb_id? Is backward compatibility important here?

File: fusedrug/data/protein/structure/structure_io.py [hunk 5]
```diff
@@ -174,7 +176,7 @@ def load_protein_structure_features(
     chain_id_type: str = "author_assigned",
     device: str = "cpu",
     max_allowed_file_size_mbs: float = None,
-    also_return_mmcif_object: bool = False,
+    # also_return_mmcif_object: bool = False,
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/structure_io.py:179]: delete?

File: fusedrug/data/protein/structure/structure_io.py [hunk 6]
```diff
@@ -694,41 +758,60 @@ def flexible_save_pdb_file(
...
     for i, (aa_idx, p_res, b, m_res) in enumerate(
-        zip(sequence, xyz, b_factors, residues_mask.bool())
+        zip(sequence, xyz, b_factors, residues_mask)
     ):
         if not m_res:
             continue
         aa_idx = aa_idx.item()
-        p_res = p_res.clone().detach().cpu()  # fixme: this looks slow
-        if aa_idx == 21:
+        # if torch.is_tensor(p_res):
```
Comment (mosheraboh) [inline fusedrug/data/protein/structure/structure_io.py:788]: delete?

**Review by mosheraboh — APPROVED** [review]
Looks great.
Minors inline.

File: fusedrug/data/protein/structure/extract_chains_to_pdbs.py (hunk 1 above)
Comment (YoelShoshan) [reply fusedrug/data/protein/structure/extract_chains_to_pdbs.py:9]: added this desc in the docstring:

"Takes an input PDB files and splits it into separate files, one per describe chain, allowing to rename the chains if desired"

File: fusedrug/data/protein/structure/flexible_align_chains_structure.py (hunk 2 above)
Comment (YoelShoshan) [reply fusedrug/data/protein/structure/flexible_align_chains_structure.py:176]: deleted

File: fusedrug/data/protein/structure/protein_complex.py (hunk 3 above)
Comment (YoelShoshan) [reply fusedrug/data/protein/structure/protein_complex.py:26]: a key is a tuple in the format (pdb_id, chain_id)

added a comment above it with description

File: fusedrug/data/protein/structure/protein_complex.py (hunk 4 above)
Comment (YoelShoshan) [reply fusedrug/data/protein/structure/protein_complex.py:33]: pdb_id_or_filename might be a path to a local pdb file, so pdb_id gives the user an option to define the pdb_id as there is no implemented way to automatically extract it.

I'll check ways to extract it automatically from the pdb file itself, and check if it's important to allow the user to override that.

That will be a separate PR.

about what you asked - backward compatibility isn't super critical here, as we can likely detect all usage locations and port them to newer usage.

File: fusedrug/data/protein/structure/structure_io.py (hunk 5 above)
Comment (YoelShoshan) [reply fusedrug/data/protein/structure/structure_io.py:179]: deleted

**Pull request merged**